orm/serial_bucket.go

```diff
@@ -0,0 +1,384 @@
+			return nil, errors.Wrap(err, "reverse prefix scan")
+		}
+	} else {
+		rawIter, err = db.Iterator(start, end)
+		if err != nil {
+			return nil, errors.Wrap(err, "prefix scan")
+		}
+	}
+
+	return &idSerialModelIterator{iterator: rawIter, bucketPrefix: smb.b.DBKey(nil)}, nil
+}
+
+// getIndexInfo returns the indexInfo from indices by name
+func (smb *serialModelBucket) getIndexInfo(name string) *indexInfo {
+	for _, info := range smb.indices {
+		if info.name == name {
+			return &info
+		}
+	}
+	return nil
+}
+
+func (smb *serialModelBucket) IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error) {
+	// get index
+	info := smb.getIndexInfo(indexName)
+	if info == nil {
+		return nil, errors.Wrapf(errors.ErrDatabase, "no index with name %s", indexName)
+	}
+
+	dbPrefix := append(info.prefix, prefix...)
+	start, end := prefixRange(dbPrefix)
+
+	var rawIter weave.Iterator
+	var err error
+	if reverse {
+		rawIter, err = db.ReverseIterator(start, end)
+		if err != nil {
+			return nil, errors.Wrap(err, "reverse prefix scan")
+		}
+	} else {
+		rawIter, err = db.Iterator(start, end)
+		if err != nil {
+			return nil, errors.Wrap(err, "prefix scan")
+		}
+	}
+
+	return &indexSerialModelIterator{
+		iterator:     rawIter,
+		bucketPrefix: smb.b.DBKey(nil),
+		unique:       info.unique,
+		kv:           db,
+	}, nil
+}
+
+func (smb *serialModelBucket) ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, destination SerialModelSlicePtr) error {
+	objs, err := smb.b.GetIndexed(db, indexName, key)
+	if err != nil {
+		return err
+	}
+	if len(objs) == 0 {
+		return nil
+	}
+
+	dest := reflect.ValueOf(destination)
+	if dest.Kind() != reflect.Ptr {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of SerialModels")
+	}
+	if dest.IsNil() {
+		return errors.Wrap(errors.ErrImmutable, "got nil pointer")
+	}
+	dest = dest.Elem()
+	if dest.Kind() != reflect.Slice {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of SerialModels")
+	}
+
+	// It is allowed to pass destination as both []MySerialModel and []*MySerialModel
+	sliceOfPointers := dest.Type().Elem().Kind() == reflect.Ptr
+
+	allowed := dest.Type().Elem()
+	if sliceOfPointers {
+		allowed = allowed.Elem()
+	}
+	if smb.model != allowed {
+		return errors.Wrapf(errors.ErrType, "this bucket operates on %s serialmodel and cannot return %s", smb.model, allowed)
+	}
+
+	for _, obj := range objs {
+		if obj == nil || obj.Value() == nil {
+			continue
+		}
+
+		val := reflect.ValueOf(obj.Value())
+		if err := val.Interface().(SerialModel).SetID(obj.Key()); err != nil {
+			return errors.Wrap(errors.ErrType, "cannot set ID")
+		}
+
+		if !sliceOfPointers {
+			val = val.Elem()
+		}
+		// store the key on the SerialModel
+		dest.Set(reflect.Append(dest, val))
+	}
+	return nil
+
+}
+
+func (smb *serialModelBucket) Create(db weave.KVStore, m SerialModel) error {
+	mTp := reflect.TypeOf(m)
```

Comment:
I see these reflect checks (7+ lines) in Create and Upsert, and something similar in the load code.

Maybe nicer to pull these out into a named function with all the reflect/type stuff and just use the name here (`enforceType`). It makes the main logic easier to read, as I know many people stop reading when they see `reflect` in go